### ISSUE TITLE
Back end - Audit submission access page

### ISSUE THREAD

**Issue by petermarks-gsa** [GSA-TTS/FAC#246]
**Background**

From parent issue #128:

In the future, non-federal users will have to log in via login.gov.

It covers the following stories:

1. As an auditor or auditee, I need to be able to specify who will need to certify my new audit (who will "check the box" at the end)
2. As an auditor or auditee, I need to be able to specify who should have access to my new audit
3. As a user, I need information on what will happen when I enter email addresses here so that I feel confident moving to the next step. (this is a front end/content change for now--not anything having to do with the mechanics of notifications)
4. As an auditor or auditee creating a new audit submission, I need the submission access page to make it clear what a “contact” is and what kind of access to the submission they’ll have so that I can give the right access to the right people.

**Acceptance criteria:**
* [x] ~~Create an API endpoint~~ Modify the `/access` API endpoint to accept information from audit submission access page
* [x] When data is submitted, a row in the single audit checklist table is created with information from the three steps of this form
* [x] When a SAC is created, we can see in Django admin that the email addresses associated with the SAC are visible in the view of `Access` entries. (If this requires extra work, it should be its own issue.)
* [x] Tests cover all changes/additions 

**Tasks:**
- [X] Expand SAC data model for audit submission access page information
- [X] ~~Create API endpoint~~ Modify the `/access` API endpoint for step 3 data
- [x] Take data stored in user profile and store in SAC 
- [X] Determine what happens to users
- [X] Add `Access`/`User` model changes in Django admin interface
- [x] Upon save, add `User`s who don't have a login.gov account (in progress)
- [x] Create new tests and ensure tests for modified things are also modified
- [x] Save data from steps to model (in progress)
  * Includes who has access 

**Related, but separate issues:**
* ~~Update the "Access" model to allow multiple auditees and auditors, but one-and-only-one certifying auditee and one-and-only-one certifying auditor. ~~ (Had to be done in this ticket)
  * ~~confirm we want only one per user research~~ confirmed by @MelissaBraxton 2022-06-23
* Endpoint that lets a user create other users for their SAC outside of the initial submission UI
  * Helpful for testing login.gov flow:
  * Without this, we have to create an entire SAC to test the creation of new users
* Sending an invite
* Display logic of whether to prevent invited email or account email of person invited.
* #320 
   * i.e. Allow auditors/auditees to be matched with existing login.gov users
* #333 
   * i.e. Allow edits of names/emails of auditors/auditees edit previous and new accesses to the SAC
* ~~#334~~
   * Issue 334 is no longer needed as the evolution of this issue used more built-in validation.

**Comment by petermarks-gsa** [comment]
This is blocked by https://github.com/GSA-TTS/FAC/issues/168

**Comment by petermarks-gsa** [comment]
This is now unblocked

**Comment by tadhg-ohiggins** [comment]
@geekygirlsarah note that `auditee_cert` and `auditor_cert` are limited to a max of one user per SAC, but `auditee_contact` and `auditee_cert` can have any number of users. I don't think there's anything in the code currently reflecthing this. (Via @MelissaBraxton )

**Comment by geekygirlsarah** [comment]
To document what was decided between @tadhg-ohiggins and @timoballard and myself today:

The `Access` model had a user (ID/email), a role (auditor/auditee and contact/certifier), and a `SingleAuditChecklist`. A `SingleAuditChecklist` had a single `Access` for each of auditor/auditee contact/certifier.

Since SACs actually need multiple peoples as auditee/auditor contacts (but still only one auditor/auditee certifier), this has to get expanded. Expanding it in the SAC (via `ManyToMany` relationship) caused circular references (SAC->Access->SAC->Access->...). So I originally out the SAC in the `Access` model.

We realized in our meeting that if we went into the admin panel to try to find out how many SACs a user was tied to, that'd be complicated as it'd require multiple steps (Look up the user's accesses, then look up all their lists). But changing the models to make that easy meant saving the multiple contacts on the front end then required multiple steps for each user (as you'd have to look up each user, then add a checklist to their access list).

Our final decision was to continue development as I had started, which was remove the SAC from `Access`. This allows the roles to still be individually identified for each SAC they're tied to, and also allow one to be added/removed from a checklist easily without making saving difficult. The admin can do a more complicated look up.

We also decided I would spend some time researching how to do that in the admin interface (as I haven't used Django's admin interface much and don't know how to do multi-layered lookups in there).

**Comment by tadhg-ohiggins** [comment]
## 2022-07-08 discussion on some edge cases around login.gov users and `Access` objects.
Participants: @geekygirlsarah @tadhg-ohiggins @timoballard 

1.  `SingleAuditChecklist` object is created with `Access` objects that have null `user_id` fields but they have the email addresses the user submitted for auditee, etc.
1.  We do not check the email addresses in the newly-created `Access` objects to see if there are `User` objects with matching email addresses—we only check for that during the login.gov login flow.
1.  When a user logs in with login.gov, we:
    1.  See if there is already a `User` for that login.gov identifier. If not, we create that `User` with their login.gov primary email address as their email address property on `User`.
    1.  Look at all their login.gov email addresses and at all `Access` objects with `null` `user_id` fields to see if any of the email addresses match.
    1.  If any match, we associate those `Access` objects with their `user_id`.

## Scenario 1

*   Alice adds bob@agency1.gov to SAC 1. `Access` object is created with bob@agency1.gov as email address.
*   Chris adds fac-liason@agency2.gov to SAC 2. `Access` object is created with fac-liaison@agency2.gov as email address.
*   Bob logs into FAC for the first time, `all_emails` contains `[bob@agency1.gov, fac-liaison@agency2.gov]`
*   We create `User` for Bob with `bobprimaryemail@agency.gov` as the canonical email address.
*   We pull up all `Access` objects with null `user_id` fields and find two that match some of Bob's entries in the login.gov `all_emails` property: SAC 1 with bob@agency1.gov and SAC 2 with fac-liaison@agency2.gov.
*   We associate the Bob `User` with both of those `Access` objects.
*   We leave the email addresses as they are in both `Access` objects even though it doesn't match the email address in the Bob `User` object.

## Scenario 2: how do we remove users from `Access` objects given the potential mismatch between email address and `User` object and login.gov emails?

*   We need to look for changes to the contact fields in the `SingleAuditChecklist` object that contain `Access` references and delete potential dangling `Access` objects that are no longer referred to by any `SingleAuditChecklist` object.
*   We look up the `Access` object by email address and delete it; we also set the corresponding property on the `SingleAuditChecklist` object to `null`.

## Scenario 3

Bob quits Agency 2. Agency 2 promotes Jane to be the new FAC Liaison. Jane adds fac-liaison@agency2.gov to her login.gov account. Bob removes fac-liaison@agency2.gov from his login.gov account (hopefully in reverse order…).

Jane logs in to FAC:

*   We create `User` for Jane with `janeprimary@agency.gov` as the email address.
*   We don't find any `Access` objects with null `user_id` fields that match any of Jane's `all_email` entries.
*   ??? profit

Tadhg proposal: the FAC users at Agency 2 need to remove fac-liaison@agency.gov from the set of allowed users and then re-add it.

New issues:
*   We need to look for changes to the contact fields in the `SingleAuditChecklist` object that contain `Access` references and delete potential dangling `Access` objects that are no longer referred to by any `SingleAuditChecklist` object.
*   We look up the `Access` object by email address and delete it; we also set the corresponding property on the `SingleAuditChecklist` object to `null`.
*   Check assumption that email address can only exist in one login.gov account at a time.
*   Find out how (if it's possible at all) email addresses may be able to migrate from one login.gov account to another.

**Comment by petermarks-gsa** [comment]
@tadhg-ohiggins Per our conversation, the first two "new issues" you will be creating new tasks for, at which point we can remove the needs refinement label. The last two are covered by #320

**Comment by petermarks-gsa** [comment]
> @tadhg-ohiggins Per our conversation, the first two "new issues" you will be creating new tasks for, at which point we can remove the needs refinement label. The last two are covered by https://github.com/GSA-TTS/FAC/issues/320

@tadhg-ohiggins Following up on our conversation last week it looks like we've created a couple additional tasks for this (#333 and #334). Are we good to remove "needs refinement"?
